guppy/src/graph/cargo.rs

```diff
@@ -0,0 +1,365 @@
+    ///
+    /// This represents the packages and features that are included as code in the final build
+    /// artifacts. This is relevant for both cross-compilation and auditing.
+    pub fn target_features(&self) -> &FeatureSet<'g> {
+        &self.target_features
+    }
+
+    /// Returns the feature set enabled on the host platform.
+    ///
+    /// This represents the packages and features that influence the final build artifacts, but
+    /// whose code is generally not directly included.
+    pub fn host_features(&self) -> &FeatureSet<'g> {
+        &self.host_features
+    }
+
+    /// Returns `PackageLink` instances for procedural macro dependencies from target packages.
+    ///
+    /// Procedural macros straddle the line between target and host: they're built for the host
+    /// but generate code that is compiled for the target platform.
+    ///
+    /// ## Notes
+    ///
+    /// Procedural macro packages will be included in the *host* feature set.
+    ///
+    /// The returned iterator will include proc macros that are depended on normally or in dev
+    /// builds from initials (if `include_dev` is set), but not the ones in the
+    /// `[build-dependencies]` section.
+    pub fn proc_macro_links<'a>(
+        &'a self,
+    ) -> impl Iterator<Item = PackageLink<'g>> + ExactSizeIterator + 'a {
+        let package_graph = self.target_features.graph().package_graph;
+        self.proc_macro_edge_ixs.iter().map(move |edge_ix| {
+            let (source_ix, target_ix) = package_graph
+                .dep_graph
+                .edge_endpoints(*edge_ix)
+                .expect("valid edge ix");
+            package_graph.edge_to_link(source_ix, target_ix, *edge_ix, None)
+        })
+    }
+}
+
+struct CargoSetBuildState<'a> {
+    opts: &'a CargoOptions<'a>,
+    omitted_packages: SortedSet<NodeIndex<PackageIx>>,
+}
+
+impl<'a> CargoSetBuildState<'a> {
+    fn new(query: &FeatureQuery<'_>, opts: &'a CargoOptions<'a>) -> Result<Self, Error> {
+        if query.direction() == DependencyDirection::Reverse {
+            return Err(Error::CargoSetError(
+                "attempted to compute for reverse query".into(),
+            ));
+        }
+
+        let omitted_packages: SortedSet<_> = query
+            .graph()
+            .package_graph
+            .package_ixs(opts.omitted_packages.iter().copied())?;
+
+        Ok(Self {
+            opts,
+            omitted_packages,
+        })
+    }
+
+    fn build(self, query: FeatureQuery) -> CargoSet {
+        match self.opts.version {
+            CargoResolverVersion::V1 => self.new_v1(query),
+        }
+    }
+
+    fn new_v1(self, query: FeatureQuery) -> CargoSet {
+        // Prepare a package query for step 2.
+        let graph = *query.graph();
+        let package_ixs: SortedSet<_> = query
+            .params
+            .initials()
+            .iter()
+            .map(|feature_ix| graph.package_ix_for_feature_ix(*feature_ix))
```

Comment:
Is `ix` and index? I think this is the first time I haven't seen it written `idx`

Reply:
Yeah, it's the convention I use in guppy to mark graph indexes.

Reply:
I think it came from petgraph, since that is how it abbreviates index as well.